File: political-podcast-corpus/r/export_regime_tokens.R
#!/usr/bin/env Rscript
# Build one curated regime's full-corpus token export for the Python LDA
# refit (nlp/lda_regime_refit.py). Reads the raw passage text (same
# boundaries as the existing pipeline, no re-segmentation), applies
# r/preprocess.R's toggles for the named regime, and writes:
#   data/output/regimes/tokens_<regime>.csv   (chunk_id, collection_id, clean_text)
#   data/output/regimes/vocab_accounting_<regime>.json
#
# Usage: Rscript r/export_regime_tokens.R <regime_name>
#   regime_name in {minimal, moderate, aggressive, aggressive_pos, stem_comparison}

.libPaths(c("r/lib", .libPaths()))
suppressMessages({
  library(quanteda)
  library(jsonlite)
})
source("r/preprocess.R")

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) stop("usage: Rscript r/export_regime_tokens.R <regime_name>")
regime_name <- args[1]

OUT_DIR <- "data/output/regimes"
dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)

CHUNKS_PATH <- "data/output/chunks_500_nolemma.csv"
CACHE_DIR <- "data/output/regimes/udpipe_cache"
dir.create(CACHE_DIR, showWarnings = FALSE, recursive = TRUE)

cat(sprintf("[%s] loading raw passages from %s...\n", regime_name, CHUNKS_PATH))
raw <- read.csv(CHUNKS_PATH, stringsAsFactors = FALSE,
                 colClasses = c(chunk_id = "character", collection_id = "character"))
raw <- raw[, c("chunk_id", "collection_id", "text")]
raw$text[is.na(raw$text)] <- ""
names(raw)[names(raw) == "chunk_id"] <- "doc_id"
cat(sprintf("[%s] %d passages loaded\n", regime_name, nrow(raw)))

toggles <- regime_toggles(regime_name)
needs_udpipe <- isTRUE(toggles$pos_filter) || isTRUE(toggles$lemmatize)

udmodel <- NULL
if (needs_udpipe) {
  source("r/udpipe_annotate_parallel.R")
  ann_cache <- file.path(CACHE_DIR, "full_corpus_annotated.rds")
  ann <- get_or_build_annotation(raw, ann_cache)
  toks <- tokens_from_annotation(ann, raw$doc_id, toggles)
} else {
  corpus_df <- raw[, c("doc_id", "text")]
  toks <- build_tokens(corpus_df, toggles, verbose = TRUE)
}

d <- dfm(toks)
if (isTRUE(toggles$infrequent_terms)) {
  d <- apply_docfreq_prune(d, verbose = TRUE)
} else {
  attr(d, "vocab_accounting") <- list(
    vocab_before = nfeat(d), vocab_after = nfeat(d),
    tokens_before = sum(d), tokens_after = sum(d),
    min_docfreq = NA, max_docfreq_prop = NA
  )
}

kept_vocab <- featnames(d)
toks_pruned <- tokens_select(toks, pattern = kept_vocab, selection = "keep")

clean_text <- vapply(as.list(toks_pruned), paste, character(1), collapse = " ")
out_df <- data.frame(
  chunk_id = raw$doc_id,
  collection_id = raw$collection_id,
  clean_text = clean_text[raw$doc_id],
  stringsAsFactors = FALSE
)
out_path <- file.path(OUT_DIR, sprintf("tokens_%s.csv", regime_name))
write.csv(out_df, out_path, row.names = FALSE)
cat(sprintf("[%s] wrote %s (%d rows, %d empty)\n", regime_name, out_path,
            nrow(out_df), sum(out_df$clean_text == "")))

acc <- attr(d, "vocab_accounting")
acc$regime <- regime_name
acc$n_docs <- nrow(out_df)
acc$n_empty_docs <- sum(out_df$clean_text == "")
acc$toggles <- toggles
write_json(acc, file.path(OUT_DIR, sprintf("vocab_accounting_%s.json", regime_name)),
           auto_unbox = TRUE, pretty = TRUE)
cat(sprintf("[%s] vocab accounting: %d -> %d vocab, %d -> %d tokens\n",
            regime_name, acc$vocab_before, acc$vocab_after, acc$tokens_before, acc$tokens_after))
cat(sprintf("[%s] DONE\n", regime_name))

File: political-podcast-corpus/nlp/export_regime_tokens.py
"""Build one curated regime's full-corpus token export for the LDA refit
(nlp/lda_regime_refit.py). Reads raw passage text (chunks_500_nolemma.csv's
`text` column -- boilerplate-sentence-stripped but otherwise unprocessed,
SAME passage boundaries as the existing pipeline, no re-segmentation),
applies nlp/preprocess_regimes.py's toggles, and writes:

    data/output/regimes/tokens_<regime>.csv         (chunk_id, collection_id, clean_text)
    data/output/regimes/vocab_accounting_<regime>.json

`aggressive` and `aggressive_pos` share ONE spaCy annotation pass (cached
to data/output/regimes/spacy_annotation_cache.pkl) since they differ only
in downstream filtering of the same tagged tokens, not in what gets
tagged -- this halves the (expensive) spaCy cost for the two POS-dependent
regimes.

Usage:
    .venv/bin/python -m nlp.export_regime_tokens --regime moderate
    .venv/bin/python -m nlp.export_regime_tokens --all
"""
from __future__ import annotations

import argparse
import json
import pickle
from pathlib import Path

import pandas as pd

from . import preprocess_regimes as pp

CHUNKS_PATH = Path("data/output/chunks_500_nolemma.csv")
OUT_DIR = Path("data/output/regimes")
SPACY_CACHE = OUT_DIR / "spacy_annotation_cache.pkl"
N_PROCESS = 5  # middle ground: fewer workers than the original 10 (~30GB peak) but faster than 3 (~10-15GB peak)


def load_raw(limit: int | None = None) -> pd.DataFrame:
    df = pd.read_csv(CHUNKS_PATH, usecols=["chunk_id", "collection_id", "text"],
                      dtype={"chunk_id": str, "collection_id": str}, nrows=limit)
    df["text"] = df["text"].fillna("")
    return df


def get_or_build_spacy_annotation(texts: list[str]) -> list:
    if SPACY_CACHE.exists():
        print(f"[cache] loading spaCy annotation from {SPACY_CACHE}")
        with open(SPACY_CACHE, "rb") as f:
            return pickle.load(f)
    print(f"[cache] no cache found, annotating {len(texts)} passages with spaCy "
          f"(n_process={N_PROCESS}) -- shared by aggressive + aggressive_pos...")
    ann = pp.annotate_spacy(texts, disfluency=True, n_process=N_PROCESS, batch_size=50, verbose=True)
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    with open(SPACY_CACHE, "wb") as f:
        pickle.dump(ann, f)
    print(f"[cache] wrote {SPACY_CACHE}")
    return ann


def export_regime(regime: str, raw: pd.DataFrame) -> None:
    toggles = pp.regime_toggles(regime)
    texts = raw["text"].tolist()

    needs_spacy = toggles.lemmatize or toggles.pos_filter
    if needs_spacy:
        ann = get_or_build_spacy_annotation(texts)
        token_lists = pp.tokens_from_annotation(ann, toggles)
        if toggles.disfluency:
            token_lists = [[t for t in toks if t not in pp.BACKCHANNEL] for toks in token_lists]
        if toggles.stopwords:
            sw = pp.extended_stopwords()
            token_lists = [[t for t in toks if t not in sw] for toks in token_lists]
        if toggles.ngrams:
            token_lists = pp.compound_collocations(token_lists, verbose=True)
    else:
        token_lists = pp.build_token_lists(texts, toggles, n_process=1, verbose=True)

    pruned, acc = pp.apply_docfreq_prune(token_lists, toggles, verbose=True)

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    out_df = pd.DataFrame({
        "chunk_id": raw["chunk_id"],
        "collection_id": raw["collection_id"],
        "clean_text": [" ".join(t) for t in pruned],
    })
    out_path = OUT_DIR / f"tokens_{regime}.csv"
    out_df.to_csv(out_path, index=False)
    n_empty = (out_df["clean_text"] == "").sum()
    print(f"[{regime}] wrote {out_path} ({len(out_df)} rows, {n_empty} empty)")

    acc["regime"] = regime
    acc["n_empty_docs"] = int(n_empty)
    acc["toggles"] = vars(toggles)
    acc_path = OUT_DIR / f"vocab_accounting_{regime}.json"
    acc_path.write_text(json.dumps(acc, indent=2, default=str))
    print(f"[{regime}] vocab accounting -> {acc_path}")
    print(f"[{regime}] DONE: vocab {acc['vocab_before']} -> {acc['vocab_after']}, "
          f"tokens {acc['tokens_before']} -> {acc['tokens_after']}")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--regime", type=str, default=None)
    ap.add_argument("--all", action="store_true")
    ap.add_argument("--limit", type=int, default=None, help="smoke-test on first N rows")
    args = ap.parse_args()

    raw = load_raw(limit=args.limit)
    print(f"[data] {len(raw)} passages loaded from {CHUNKS_PATH}")

    if args.all:
        # spaCy-needing regimes first so the shared cache is built once
        regimes = ["aggressive", "aggressive_pos", "minimal", "moderate"]
    elif args.regime:
        regimes = [args.regime]
    else:
        ap.error("pass --regime NAME or --all")
        return

    for r in regimes:
        print(f"\n{'=' * 70}\n[export] regime = {r}\n{'=' * 70}")
        export_regime(r, raw)


if __name__ == "__main__":
    main()
